HW3/P4/median_filter.cl

```diff
@@ -31,4 +31,72 @@ median_3x3(__global __read_only float *in_values,
 
     // Each thread in the valid region (x < w, y < h) should write
     // back its 3x3 neighborhood median.
+    
+    // Global position of output pixel
+    const int x = get_global_id(0);
+    const int y = get_global_id(1);
+
+    // Local position relative to (0, 0) in workgroup
+    const int lx = get_local_id(0);
+    const int ly = get_local_id(1);
+
+    // coordinates of the upper left corner of the buffer in image
+    // space, including halo
+    const int buf_corner_x = x - lx - halo;
+    const int buf_corner_y = y - ly - halo;
+
+    // coordinates of our pixel in the local buffer
+    const int buf_x = lx + halo;
+    const int buf_y = ly + halo;
+
+    // 1D index of thread within our work-group
+    const int idx_1D = ly * get_local_size(0) + lx;
+
+    int row;
+
+    // Since the kernels are in order, by loading by column per kernel, 
+    // we'll actually be loading by row across kernels
+    if (idx_1D < buf_w) {
+        for (row = 0; row < buf_h; row++) {
+            int tmp_x = idx_1D;
+            int tmp_y = row;
+            
+            if (buf_corner_x + tmp_x < 0) {
+                tmp_x++;
+            } else if (buf_corner_x + tmp_x >= w) {
+                tmp_x--;
```

Comment:
This may cause an out-of-bounds access to the original input data, because the buffer on the right-most work group might have more than 1 column to the right of the input matrix.